src/core/src/ffi/storage.rs

```diff
@@ -0,0 +1,140 @@
+use std::os::raw::c_char;
+use std::slice;
+
+use crate::ffi::utils::{ForeignObject, SourmashStr};
+use crate::prelude::*;
+use crate::storage::ZipStorage;
+
+pub struct SourmashZipStorage;
+
+impl ForeignObject for SourmashZipStorage {
+    type RustObject = ZipStorage;
+}
+
+ffi_fn! {
+unsafe fn zipstorage_new(ptr: *const c_char, insize: usize) -> Result<*mut SourmashZipStorage> {
+    let path = {
+        assert!(!ptr.is_null());
+        let path = slice::from_raw_parts(ptr as *mut u8, insize);
+        std::str::from_utf8(path)?
+    };
+    let zipstorage = ZipStorage::from_file(path)?;
+
+    Ok(SourmashZipStorage::from_rust(zipstorage))
+}
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn zipstorage_free(ptr: *mut SourmashZipStorage) {
+    SourmashZipStorage::drop(ptr);
+}
+
+ffi_fn! {
+unsafe fn zipstorage_load(ptr: *const SourmashZipStorage,
+    path_ptr: *const c_char,
+    insize: usize,
+    size: *mut usize) -> Result<*const u8> {
+
+    let storage = SourmashZipStorage::as_rust(ptr);
+
+    let path = {
+        assert!(!path_ptr.is_null());
+        let path = slice::from_raw_parts(path_ptr as *mut u8, insize);
+        std::str::from_utf8(path)?
+    };
+
+    let buffer = storage.load(path)?;
+
+    let b = buffer.into_boxed_slice();
+    *size = b.len();
+
+    Ok(Box::into_raw(b) as *const u8)
+}
+}
+
+ffi_fn! {
+unsafe fn zipstorage_list_sbts(
+    ptr: *const SourmashZipStorage,
+    size: *mut usize,
+) -> Result<*mut *mut SourmashStr> {
+    let storage = SourmashZipStorage::as_rust(ptr);
+
+    let sbts = storage.list_sbts()?;
+
+    // FIXME: use the ForeignObject trait, maybe define new method there...
+    let ptr_sigs: Vec<*mut SourmashStr> = sbts
+        .into_iter()
+        .map(|x| Box::into_raw(Box::new(SourmashStr::from_string(x))) as *mut SourmashStr)
+        .collect();
+
+    let b = ptr_sigs.into_boxed_slice();
+    *size = b.len();
+
+    Ok(Box::into_raw(b) as *mut *mut SourmashStr)
+}
+}
+
+ffi_fn! {
+unsafe fn zipstorage_filenames(
+    ptr: *const SourmashZipStorage,
+    size: *mut usize,
+) -> Result<*mut *mut SourmashStr> {
+    let storage = SourmashZipStorage::as_rust(ptr);
+
+    let files = storage.filenames()?;
+
+    // FIXME: use the ForeignObject trait, maybe define new method there...
```

Comment:
Would really appreciate elaborating more on the FIXME/TODO messages, like why? or a little hint :)

Reply:
This function is pretty much copy-and-pasted from https://github.com/sourmash-bio/sourmash/blob/1229dc1d6664004e580e3b80d4c13a8ae12a19be/src/core/src/ffi/signature.rs#L284L292

`ForeignObject` is defined [here](https://github.com/sourmash-bio/sourmash/blob/1229dc1d6664004e580e3b80d4c13a8ae12a19be/src/core/src/ffi/utils.rs#L21L56), and collect some common functionality for working with the FFI layer (going from a Rust object to a raw pointer, or from a raw pointer to a Rust object). What is missing there is working with arrays of objects (like this function here, which returns a vector of `SourmashStr`).

Most of the decisions/designs in the FFI layer in sourmash comes from [symbolic](https://github.com/getsentry/symbolic/blob/d938048ccf5df9cd7cd41afdc419eb5a92ff7b84/symbolic-cabi/src/utils.rs), and you can see other ways to use `ForeignObject` over there)